trin-core/src/utp/packets.rs

```diff
@@ -0,0 +1,322 @@
+    while extension != 0 {
+        if data.len() < index + 2 {
+            return Err("Invalid packet length");
+        }
+        extension = data[index];
+        let len = data[index + 1] as usize;
+
+        if len == 0 || len % 4 != 0 || len + index + 2 > data.len() {
+            return Err("Invalid Extension Length");
+        }
+
+        index += len + 2;
+    }
+    Ok(())
+}
+
+impl<'a> TryFrom<&'a [u8]> for Packet {
+    type Error = &'static str;
+
+    // todo: Added support to check validity of extensions,
+    // we don't need to worry about checking payloads
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        match PacketHeader::try_from(value).and(check_extensions(value)) {
+            Ok(_) => Ok(Packet(value.to_owned())),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl Default for PacketHeader {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub struct PacketHeader {
+    type_ver: u8,
+    extension: u8,
+    connection_id: u16,
+    // This time is in microseconds
+    timestamp: u32,
+    timestamp_difference: u32,
+    wnd_size: u32,
+    pub seq_nr: u16,
+    ack_nr: u16,
+}
+
+impl PacketHeader {
+    pub fn new() -> Self {
+        PacketHeader {
+            type_ver: u8::from(PacketType::Data) << 4 | VERSION,
+            extension: 0,
+            connection_id: 0,
+            timestamp: 0,
+            timestamp_difference: 0,
+            wnd_size: 0xf000,
+            seq_nr: 0,
+            ack_nr: 0,
+        }
+    }
+
+    pub fn encode(&mut self) -> Vec<u8> {
+        let mut buf = vec![];
+        buf.extend_from_slice(&[self.type_ver]);
+        buf.extend_from_slice(&[self.extension]);
+        buf.extend_from_slice(&self.connection_id.to_be_bytes());
+        buf.extend_from_slice(&self.timestamp.to_be_bytes());
+        buf.extend_from_slice(&self.timestamp_difference.to_be_bytes());
+        buf.extend_from_slice(&self.wnd_size.to_be_bytes());
+        buf.extend_from_slice(&self.seq_nr.to_be_bytes());
+        buf.extend_from_slice(&self.ack_nr.to_be_bytes());
+        buf
+    }
+
+    pub fn decode(bytes: &[u8]) -> PacketHeader {
+        PacketHeader {
+            type_ver: bytes[0],
+            extension: bytes[1],
+            connection_id: u16::from_be_bytes([bytes[2], bytes[3]]),
+            timestamp: u32::from_be_bytes([bytes[4], bytes[5], bytes[6], bytes[7]]),
+            timestamp_difference: u32::from_be_bytes([bytes[8], bytes[9], bytes[10], bytes[11]]),
+            wnd_size: u32::from_be_bytes([bytes[12], bytes[13], bytes[14], bytes[15]]),
+            seq_nr: u16::from_be_bytes([bytes[16], bytes[17]]),
+            ack_nr: u16::from_be_bytes([bytes[18], bytes[19]]),
+        }
+    }
+
+    pub fn version(&self) -> u8 {
+        self.type_ver & 0xf
+    }
+    pub fn set_version(&mut self, int: u8) {
+        self.type_ver = (self.type_ver & 0xf0) | (int & 0xf)
```

Comment:
This is really confusing for me, in this method we are setting a private field `self.type_var` and below we have another method that is also setting the same field:

```rust
pub fn set_type(&mut self, t: PacketType) {
        self.type_ver = (self.type_ver & 0xf) | (u8::from(t) << 4)
    }
```
So we have two setters for the same field, what is the case here when to use one over another?

Reply:
Please refer to the spec
![image](https://user-images.githubusercontent.com/31669092/150379771-812a763d-a50e-465b-b3fc-99fb0ae2aa57.png)

So the smallest int literal is a byte or u8, but it would be wasteful over the wire to have 2 separate bytes for such small values so we split the u8 in half 0000_0000 so the first half will store the type and the second half will store the version of the packet.

If you look at set_type.  there is a 4-bit shift when we right to it and when we use set_version there isn't.

Reply:
Do you think this would be an acceptable comment for that?

![image](https://user-images.githubusercontent.com/31669092/150383262-30597f15-1387-4398-b7c5-557177e12d2b.png)

Reply:
> So the smallest int literal is a byte or u8, but it would be wasteful over the wire to have 2 separate bytes for such small values so we split the u8 in half 0000_0000 so the first half will store the type and the second half will store the version of the packet.

I see, it would be helpful to add a comment in the code about this.

Reply:
> Do you think this would be an acceptable comment for that?

I like this more :)

```rust
/// It would be wasteful over the wire to have 2 separate bytes for type and ver, so we split the u8 in half 0000_0000 
/// so the first half will store the type and the second half will store the version of the packet
```
PS: When commenting with `///` instead of `//` this serves as [doc comments](https://doc.rust-lang.org/rust-by-example/meta/doc.html) which is used right above data structures, methods, and fields.

PS 1: I think maybe it is better if you need to post a code in github, to use a [code block](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting) as ```rust instead of uploading a code screenshot. :thinking: